src/coreComponents/fileIO/vtk/VTKPolyDataWriterInterface.cpp

```diff
@@ -296,230 +246,299 @@ VTKPolyDataWriterInterface::getVtkCells( CellElementRegion const & er ) const
+      rtTypes::applyArrayTypeLambda2( rtTypes::typeID( wrapper.getTypeId() ), true, [&]( auto array, auto scalar )
+      {
+        auto typedData = vtkAOSDataArrayTemplate< decltype( scalar ) >::New();
+        data = typedData;
+        setComponentMetadata( wrapper.cast< decltype( array ) >(), *typedData );
+      } );
 
-      nCompInSecondDim = sourceArray.size( 1 );
-      nCompInThirdDim = sourceArray.size( 2 );
-    } );
-  } );
+      data->SetNumberOfTuples( nodeManager.size() );
+      data->SetName( wrapper.getName().c_str() );
 
-  GEOSX_THROW_IF( nCompInSecondDim == 0 || nCompInThirdDim == 0,
-                  "The array4d should have at least one component in the second and third dimensions",
-                  InputError );
+      writeField( wrapper, 0, *data );
+      pointData.AddArray( data );
+    }
+  }
+}
 
-  // Step 2: we know how many components we have in the two dimensions that will be displayed component by component
-  //         we can now loop over the components of these dimensions and output to VTK in separate instances of VTKGEOSXData
-  for( localIndex iComp2 = 0; iComp2 < nCompInSecondDim; ++iComp2 )
+template< class SUBREGION >
+void VTKPolyDataWriterInterface::writeElementFields( vtkCellData & cellData,
+                                                     ElementRegionBase const & er ) const
+{
+  std::unordered_set< string > regularFields;
+  std::unordered_set< string > materialFields;
+  er.forElementSubRegions< SUBREGION >( [&]( SUBREGION const & esr )
   {
-    for( localIndex iComp3 = 0; iComp3 < nCompInThirdDim; ++iComp3 )
+    for( auto const & wrapperIter : esr.wrappers() )
     {
-      localIndex count = 0;
-
-      // Step 2.1: write the name of the specific component iComp we consider here
-      vtkSmartPointer< VTKGEOSXData > data = VTKGEOSXData::New();
-      data->SetNumberOfValues( er.getNumberOfElements< SUBREGION >() );
-
-      if( nCompInSecondDim == 1 )
-      {
-        data->SetName( (field+"_"+std::to_string( iComp3 )).c_str() ); // special case for nice display of constitutive fields
-      }
-      else
+      string const & name = wrapperIter.first;
+      auto const & wrapper = *wrapperIter.second;
+      if( wrapper.getPlotLevel() <= m_plotLevel )
       {
-        data->SetName( (field+"_"+std::to_string( iComp2 )+"_"+std::to_string( iComp3 )).c_str() );
+        // Slightly cursed: split fields into regular and material fields based on whether wrapper name begins
+        // with a known material name. I can't think of an easier way at the moment, but there ought to be one.
+        auto const startsWithMaterial = [&]( string const & mat ){ return name.substr( 0, mat.size() ) == mat; };
+        if( std::find_if( er.getMaterialList().begin(), er.getMaterialList().end(), startsWithMaterial ) != er.getMaterialList().end() )
+        {
+          materialFields.insert( name );
+        }
+        else
+        {
+          regularFields.insert( name );
+        }
```

Comment:
Ideas?

Reply:
Hmm, I'm afraid I wrote something wrong, be careful with my suggestion.

Reply:
I guess my question was indeed more on the content side: do we have a good way in our data repository structure to differentiate between "normal" fields allocated by solvers and "material" fields allocated by constitutive models? The latter have an extra gauss point dimension that needs to be handled separately. I'm doing that based on field name prefixes, but that seems super brittle. The user could name a material something that happens to be a prefix of an existing field name in one of the solvers.

Reply:
Yes, I understood that it was about content. But I think that my code snippet is wrong nevertheless. 😉

Reply:
We used to have a differentiation between regular fields and material fields....but folks weren't happy about having to reach into an extra layer for material data access, so I flattened it. I am not opposed to adding the layer back in.

Reply:
I ended up collecting material fields by iterating over constitutive models directly rather than proxy wrappers on subregions, then filtering out those names from the list of subregion wrappers to get the rest. In other words, VTK output will now also be using the extra layer of access.

I'm not sure there are any pieces of code left that use the "flattened" data model. Definitely not any physics solvers I'm aware of, nor Silo or Blueprint. I think if someone wanted to collect a time history of a material field (say stress), they would have to specify `fieldName="rock_stress"` in a `PackCollection` task and it would collect through this proxy wrapper as there is no other way currently. Other than than, we could remove the material data wrappers from subregions.